pixy/admin.go

```diff
@@ -0,0 +1,180 @@
+package pixy
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/mailgun/kafka-pixy/Godeps/_workspace/src/github.com/mailgun/sarama"
+)
+
+type (
+	ErrAdminSetup error
+)
+
+const (
+	ProtocolVer1 = 1 // Supported by Kafka v0.8.2 and later
+)
+
+type Admin struct {
+	kafkaClient sarama.Client
+}
+
+func SpawnAdmin(config *Config) (*Admin, error) {
+	kafkaClient, err := sarama.NewClient(config.Kafka.SeedPeers, config.saramaConfig())
+	if err != nil {
+		return nil, ErrAdminSetup(fmt.Errorf("failed to create sarama.Client: err=(%v)", err))
+	}
+	a := Admin{kafkaClient: kafkaClient}
+	return &a, nil
+}
+
+func (a *Admin) Stop() {
+	a.kafkaClient.Close()
+}
+
+type PartitionOffset struct {
+	Partition int32
+	Range     struct {
+		Begin int64
+		End   int64
+	}
+	Offset   int64
+	Metadata string
+}
+
+type indexedPartition struct {
+	index     int
+	partition int32
+}
+
+// GetGroupOffsets for every partition of the specified topic it returns the
+// current offset range along with the latest offset and metadata committed by
+// the specified consumer group.
+func (a *Admin) GetGroupOffsets(group, topic string) ([]PartitionOffset, error) {
+	var err error
+	partitions, err := a.kafkaClient.Partitions(topic)
+	if err != nil {
+		return nil, err
+	}
+
+	// Figure out distribution of partitions among brokers.
+	brokerToPartitions := make(map[*sarama.Broker][]indexedPartition)
+	for i, p := range partitions {
+		broker, err := a.kafkaClient.Leader(topic, p)
+		if err != nil {
+			return nil, err
+		}
+		brokerToPartitions[broker] = append(brokerToPartitions[broker], indexedPartition{i, p})
+	}
+
+	// Query brokers for the oldest and newest offsets of the partitions that
+	// they are leaders for.
+	offsets := make([]PartitionOffset, len(partitions))
+	var wg sync.WaitGroup
+	errorsCh := make(chan error, len(brokerToPartitions))
+	for broker, brokerPartitions := range brokerToPartitions {
+		broker, brokerPartitions := broker, brokerPartitions
+		var reqNewest sarama.OffsetRequest
+		var reqOldest sarama.OffsetRequest
+		for _, p := range brokerPartitions {
+			reqNewest.AddBlock(topic, p.partition, sarama.OffsetNewest, 1)
+			reqOldest.AddBlock(topic, p.partition, sarama.OffsetOldest, 1)
+		}
+		spawn(&wg, func() {
```

Comment:
That is the whole point of having [spawn](https://github.com/mailgun/kafka-pixy/blob/master/pixy/utils.go#L17)